rfcs/20230517-forward-compatibility-testing.md

```diff
@@ -0,0 +1,103 @@
+# StableHLO Forward Compatibility Testing
+
+## Background
+
+StableHLO needs to verify both forward and backward compatibility. Currently we
+check backward compatibility by versioning and serializing
+`stablehlo_legalize_to_vhlo.mlir`, and need something similar for forward
+compatibility. This document makes the following proposal:
+
+**Proposal:** Add tests to compare `stablehlo_legalize_to_vhlo.0_X_0.mlir`
+serialized at HEAD using `--target 0.X.0` with existing [`stablehlo_legalize_to_vhlo.0_X_0.mlir.bc`](https://github.com/openxla/stablehlo/blob/main/stablehlo/tests/stablehlo_legalize_to_vhlo.0_9_0.mlir.bc)
+test files.
+
+## Requirements
+
+Forward compatibility testing has the following requirements, to ensure that it
+is both useful and practical:
+
+* **R1.** Tests must detect changes in StableHLO that break forward compatibility.
+* **R2.** Tests must detect LLVM revision bumps that break forward compatibility.
+* **R3.** Tests must scale with releases and execute in a reasonable amount
+of time.
+
+## Proposed Design: Statically Test Forward Compatibility
+
+In this design, forward incompatibilities are detected by serializing
+`stablehlo_legalize_to_vhlo.0_X_0.mlir` at HEAD and comparing it to a
+known-good file serialized at previous releases. This approach
+operates on the assumption that at HEAD we are able to produce a byte-identical
+serialized artifact, excluding custom header info and debug locations.
+
+Detecting forward incompatiblities can be accomplished by adding the following
+`RUN` line to the versioned [`stablehlo_legalize_to_vhlo.0_X_0.mlir`](https://github.com/search?q=repo%3Aopenxla%2Fstablehlo+path%3A**%2Fstablehlo_legalize_to_vhlo.0_*&type=code)
+test files.
+
+```bash
+diff %s.bc <(stablehlo-translate --serialize --target=0.X.0)
+```
+
+This test should be run in CI on all PRs to prevent merging incompatible changes
+in StableHLO (_R1_) and LLVM revision bumps (_R2_).
+
+This will not work as is because `bytecode::producer`, `op::location` and
+`block_argument::location` fields may change between when the original bytecode
+file has been produced and when the test runs at HEAD. We can set a consistent
+producer string to take care of the first item. For debug locations, one option
+would be to explicitly specify debug locations, however this is very cumbersome
+in a 2.2k line test file. Another option is to strip this debug information
```

Comment:
There is a simple way: parse and print with debug info. The default behavior is to use file:line:col. That results in every location getting set to the current one.

Reply:
Ah that's a good point, actually I think we can incorporate hard-coded debug locations. Updated https://github.com/openxla/stablehlo/issues/1633 to include this with a proposed plan.